src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -296,6 +305,34 @@ internal BoundStatement BindPossibleEmbeddedStatement(StatementSyntax node, Diag
                     Debug.Assert(binder != null);
                     return binder.WrapWithVariablesIfAny(switchStatement.Expression, binder.BindStatement(node, diagnostics));
 
+                case SyntaxKind.EmptyStatement:
+                    var emptyStatement = (EmptyStatementSyntax)node;
+                    switch (node.Parent.Kind())
+                    {
+                        case SyntaxKind.ForStatement:
+                        case SyntaxKind.ForEachStatement:
+                        case SyntaxKind.ForEachVariableStatement:
+                        case SyntaxKind.WhileStatement:
+                            // For loop constructs, only warn if we see a block following the statement.
+                            // That indicates code like:  "while (x) ; { }"
+                            // which is most likely a bug.
+                            if (emptyStatement.SemicolonToken.GetNextToken().Kind() != SyntaxKind.OpenBraceToken)
+                            {
+                                break;
+                            }
+
+                            goto default;
+
+                        default:
+                            // For non-loop constructs, always warn.  This is for code like:
+                            // "if (x) ;" which is almost certainly a bug.
+                            diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation());
```

Comment:
>diagnostics.Add(ErrorCode.WRN_PossibleMistakenNullStatement, node.GetLocation()); [](start = 28, length = 81)

It feels like this error should be reported only one the semicolon token is present in code, i.e. is not missing, but it doesn't look like we are making this check. Could you please add a test for a scenario like this?